app/src/main/java/org/ea/sqrl/utils/PasswordStrengthMeter.java

```diff
@@ -0,0 +1,230 @@
+                if (count == null) count = 0;
+
+                if (count == 0) result.StrengthPoints+=1; // increase only once per class
+                result.CharClassCounts.put(characterClass, count + 1);
+
+                if (isCancelled()) return null;
+            }
+
+            if (result.StrengthPoints < STRENGTH_POINTS_MIN_MEDIUM) result.Rating = PasswordRating.Poor;
+            else if (result.StrengthPoints < STRENGTH_POINTS_MIN_GOOD) result.Rating = PasswordRating.Medium;
+            else result.Rating = PasswordRating.Good;
+
+            // Last chance to detect cancellation
+            if (isCancelled()) return null;
+
+            return result;
+        }
+
+        @Override
+        protected void onPostExecute(PasswordStrengthResult result) {
+
+            if (result == null) return;
+
+            try {
+                TextView txtPasswordStrength = mPassStrengthLayout.findViewById(R.id.txtPasswordStrength);
+                TextView txtPasswordWarning = mPassStrengthLayout.findViewById(R.id.txtPasswordWarning);
+                ImageView imgUppercase = mPassStrengthLayout.findViewById(R.id.imgPasswordContainsUppercase);
+                ImageView imgLowercase = mPassStrengthLayout.findViewById(R.id.imgPasswordContainsLowercase);
+                ImageView imgDigits = mPassStrengthLayout.findViewById(R.id.imgPasswordContainsDigits);
+                ImageView imgSymbols = mPassStrengthLayout.findViewById(R.id.imgSymbols);
+                ProgressBar progressPwStrength = mPassStrengthLayout.findViewById(R.id.progressPasswordStrength);
+
+                progressPwStrength.setMax(STRENGTH_POINTS_MIN_GOOD);
+                progressPwStrength.setProgress(result.StrengthPoints);
+
+                switch (result.Rating) {
+
+                    case Poor:
+                        txtPasswordStrength.setText(mContext.getText(R.string.password_strength_poor));
+                        txtPasswordStrength.setBackgroundColor(
+                                ContextCompat.getColor(mContext, R.color.password_strength_poor));
+                        break;
+
+                    case Medium:
+                        txtPasswordStrength.setText(mContext.getText(R.string.password_strength_medium));
+                        txtPasswordStrength.setBackgroundColor(
+                                ContextCompat.getColor(mContext, R.color.password_strength_medium));
+                        break;
+
+                    case Good:
+                        txtPasswordStrength.setText(mContext.getText(R.string.password_strength_good));
+                        txtPasswordStrength.setBackgroundColor(
+                                ContextCompat.getColor(mContext, R.color.password_strength_good));
+                        break;
+                }
+
+                ImageView iv = null;
+                for (CharacterClass cc : result.CharClassCounts.keySet()) {
```

Comment:
Again booleans would not require a loop and would only set colored drawables on the images directly. Less code and more clear.

Reply:
Agreed.